add: mplex (#269)
libp2p/docs: content/concepts/multiplex/mplex.md

---
title: "Mplex"
description: "mplex is a stream multiplexer developed for libp2p."
weight: 170
---

## What is Mplex?

> Mplex is a stream multiplexer designed for libp2p.

Reviewer: ```suggestion
mplex is a stream multiplexer designed for libp2p.
```

It is a simple protocol that does not provide many features offered by other
stream multiplexers. Notably, mplex does not provide backpressure at the protocol

> level.

Reviewer: Make a separate paragraph that lists the drawbacks:
* no backpressure on data sent on streams
* no limits on how many streams the peer can open

Author: great

Mplex runs over a reliable, ordered pipe between two peers, such as a TCP socket
or a Unix pipe. Peers can open, write to, close, and reset a stream by sending messages
with the appropriate [flag value](https://github.com/libp2p/specs/tree/master/mplex#flag-values).

> Every message (one byte) in mplex consists of a header and a length of prefixed data

Reviewer: I don't understand the "one byte" part.

---
title: "mplex"
description: "mplex is a simple stream multiplexer that was developed for libp2p."
weight: 170
---

## What is mplex?

> mplex is a stream multiplexer designed for libp2p.

Reviewer: ```suggestion
mplex is a simple stream multiplexer that was designed in the early days of libp2p.
```

It is a simple protocol that does not provide many features offered by other

> stream multiplexers. Notably, mplex does not provide backpressure at the protocol
> level.

Reviewer: ```suggestion
stream multiplexers. Notably, mplex does not provide flow control, a feature which
is now considered critical for a stream multiplexer.
```

mplex runs over a reliable, ordered pipe between two peers, such as a TCP connection.
Peers can open, write to, close, and reset a stream. mplex uses a message-based framing
layer like [yamux](yamux), enabling it to multiplex different
data streams, including stream-oriented data and other types of messages.

### Drawbacks

mplex does not have any flow control.
> Backpressure is a mechanism to prevent one peer from overwhelming a slow time consuming the data.

> mplex has no limits on how many streams a peer can open. This allows mplex to support a wide
> range of applications and use cases.

Reviewer: ```suggestion
mplex also doesn't limit how many streams a peer can open.
```

{{< alert icon="" context="">}}

> **Yamux should be used over mplex in libp2p**. It is well-suited for applications that
> require fast, reliable data transfer.

Reviewer: ```suggestion
**Yamux should be used over mplex in libp2p**. As it natively supports flow control, it is better suited for applications that require the transfer of large amounts of data.
```

---
title: "mplex"

> description: "mplex is a stream multiplexer developed for libp2p."

Reviewer: ```suggestion
description: "mplex is a simple stream multiplexer developed for libp2p."
```

weight: 170
---

## What is mplex?

mplex is a stream multiplexer designed for libp2p.
It is a simple protocol that does not provide many features offered by other
stream multiplexers. Notably, mplex does not provide backpressure at the protocol
level.

mplex runs over a reliable, ordered pipe between two peers, such as a TCP connection

> or a Unix pipe. Peers can open, write to, close, and reset a stream by sending messages

Reviewer: We don't really use Unix pipes anywhere, so it's probably not worth mentioning.

> with the appropriate [flag value](https://github.com/libp2p/specs/tree/master/mplex#flag-values).

Reviewer: It doesn't really matter in a documentation how these things are encoded.

It uses a message-based framing layer like [yamux](yamux), enabling it to multiplex different
data streams, including stream-oriented data and other types of messages.

> Every message in mplex consists of a header and the prefixed data segment length.

Reviewer: This is also probably too detailed.

### Drawbacks

mplex does not have any flow control.
> Backpressure is a mechanism to prevent one peer from overwhelming a slow time consuming the data.

mplex has no limits on how many streams a peer can open. This allows mplex to support a wide

> range of applications and use cases, but it also means that mplex may not be as efficient as other

Reviewer: Not really. QUIC has that and I wouldn't say that this limits QUIC's use cases.

multiplexing protocols that limit the number of streams that can be opened.

{{< alert icon="" context="">}}
**Yamux should be used over mplex in libp2p**. It is well-suited for applications that

> require fast, reliable data transfer as it is optimized for low-latency, high-bandwidth environments.

Reviewer: ```suggestion
require fast, reliable data transfer.
```

---
title: "mplex"
description: "mplex is a stream multiplexer developed for libp2p."
weight: 170
---

## What is mplex?

mplex is a stream multiplexer designed for libp2p.
It is a simple protocol that does not provide many features offered by other
stream multiplexers. Notably, mplex does not provide backpressure at the protocol
level.

mplex runs over a reliable, ordered pipe between two peers, such as a TCP socket
or a Unix pipe. Peers can open, write to, close, and reset a stream by sending messages
with the appropriate [flag value](https://github.com/libp2p/specs/tree/master/mplex#flag-values).

Every message in mplex consists of a header and a length of prefixed data
segment.

### Drawbacks

mplex does not support backpressure because it is designed to be more flexible and

> extensible than other multiplexing protocols. Instead of using a flow control mechanism

Reviewer: It's not really flexible and extensible. I think it was primarily designed for implementation simplicity, without really considering that backpressure is kind of essential when writing a networking stack. Of course we shouldn't put it like this in our docs ;)

> like [yamux](yamux), mplex uses a message-based framing layer that enables it to multiplex

Reviewer: Both mplex and yamux define a framing layer.

> different types of data streams, including stream-oriented data, datagrams, and other types

Reviewer: mplex doesn't have datagrams.

of messages. This flexibility comes at the cost of some performance and reliability, as

> mplex is less efficient at transferring large amounts of data than other multiplexing

Reviewer: That's not really true, if you look at raw throughput, mplex is not doing badly. It's able to utilize the full underlying TCP connection.

protocols that support backpressure.

Additionally, mplex has no limits on how many streams a peer can open. This allows mplex
to support a wide range of applications and use cases, but it also means that mplex may
not be as efficient as other multiplexing protocols that limit the number of streams that
can be opened.

{{< alert icon="" context="">}}
**Yamux should be used over mplex in libp2p**. It is well-suited for applications that

> require fast, reliable data transfer, such as real-time communication or streaming media,

Reviewer: My comment from the yamux document about real-time communication applies here too.

as it is optimized for low-latency, high-bandwidth environments.

Until recently, the reason mplex was still supported was compatibility with js-libp2p,
which didn't have yamux support.
Now that
[js-libp2p has gained yamux support](https://github.com/ChainSafe/js-libp2p-yamux/releases/tag/v1.0.0),

> you should only support mplex if you need to deal with legacy nodes.

Reviewer: ```suggestion
mplex should only be used to provide backwards-compatibility with legacy nodes.
```
